mlx_graphs/utils/transformations.py

```diff
@@ -0,0 +1,81 @@
+import mlx.core as mx
+import numpy as np
+
+
+def check_adjacency_matrix(func):
+    """Decorator function to check the validity of an adjacency matrix."""
+
+    def wrapper(adjacency_matrix, *args, **kwargs):
+        if adjacency_matrix.ndim != 2:
+            raise ValueError(
+                f"Adjacency matrix must be two-dimensional (got {adjacency_matrix.ndim} dimensions)"
+            )
+        if not mx.equal(*adjacency_matrix.shape):
+            raise ValueError(
+                f"Adjacency matrix must be a square matrix (got {adjacency_matrix.shape} shape)"
+            )
+        return func(adjacency_matrix, *args, **kwargs)
+
+    return wrapper
+
+
+@check_adjacency_matrix
+def to_edge_index(adjacency_matrix: mx.array) -> mx.array:
+    """
+    Converts an adjacency matrix to an edge index representation.
+
+    Args:
+        adjacency_matrix (mlx.core.array): the input adjacency matrix
+
+    Returns:
+        mlx.core.array: a [2, num_edges] array representing the source and target nodes of each edge
+
+    .. code-block:: python
+        matrix = mx.array(
+            [
+                [0, 1, 2],
+                [3, 0, 0],
+                [5, 1, 2],
+            ]
+        )
+        edge_index = to_edge_index(matrix)
+        # mx.array([[0, 0, 1, 2, 2, 2], [1, 2, 0, 0, 1, 2]])
+    """
+    input_dtype = adjacency_matrix.dtype
+    edge_index = mx.stack(
+        [mx.array(x, dtype=input_dtype) for x in np.nonzero(adjacency_matrix)]
+    )
+    return edge_index
+
+
+@check_adjacency_matrix
+def to_sparse_adjacency_matrix(adjacency_matrix: mx.array) -> tuple[mx.array, mx.array]:
```

Comment:
Isn't `sparse` ambiguous here? A sparse adj matrix usually refers to a specific sparse data type like a `SparseTensor` in PyG. We'll certainly implement these sparse types in the future so maybe using here `to_edge_index_with_features` would be less ambiguous?

Reply:
I'm not sure that's ambiguous here as it's pretty much how a sparse matrix is defined/represented. Also looking at PyG, it seems their [`dense_to_sparse`](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/sparse.html#dense_to_sparse) is pretty similar and doesn't necessarily imply the use of sparse tensors. The `adjacency_matrix` input may be a sparse tensor once they're implemented, but the edge index and features that this function outputs won't be sparse by definition (as they'll store information about nonzero elements of the input `adjacency_matrix`).

@TristanBilot what do you think?

Reply:
If it's implemented similarly in PyG then go for it!